platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/recovery/internal/EventStreamBound.java

```diff
@@ -0,0 +1,287 @@
+     * are strictly less, and equal otherwise.   If only one of the timestamp or round is present, the normal comparison
+     * rules apply. If neither the timestamp nor round is present, the bound is an unbounded upper bound and the event
+     * is always less than it.
+     *
+     * @param consensusData the consensus data to compare to
+     * @param boundType     the type of bound
+     * @return A positive value if the consensus data is greater than the bound, A negative value if the consensus data
+     * is less than the bound, and 0 otherwise.
+     */
+    public int compareTo(@NonNull final ConsensusData consensusData, BoundType boundType) {
+        if (isUnbounded()) {
+            return boundType == BoundType.LOWER ? 1 : -1;
+        }
+        if (round == NO_ROUND) {
+            return compareTimestamp(consensusData);
+        }
+        if (timestamp == NO_TIMESTAMP) {
+            return compareRound(consensusData);
+        }
+
+        int roundCompare = compareRound(consensusData);
+        int timestampCompare = compareTimestamp(consensusData);
+
+        if (boundType == BoundType.LOWER) {
+            if (roundCompare > 0 && timestampCompare > 0) {
+                return 1;
+            }
+            if (roundCompare < 0 || timestampCompare < 0) {
+                return -1;
+            }
+        } else {
+            if (roundCompare > 0 || timestampCompare > 0) {
+                return 1;
+            }
+            if (roundCompare < 0 && timestampCompare < 0) {
+                return -1;
+            }
+        }
+        return 0;
+    }
+
+    /**
+     * Compares the round received of the consensus data to the round of the bound.
+     *
+     * @param consensusData the consensus data to compare to
+     * @return a negative integer, zero, or a positive integer based on whether the consensus data is less, equal or greater than the bound.
+     */
+    private int compareRound(@NonNull final ConsensusData consensusData) {
+        return Long.compare(consensusData.getRoundReceived(), round);
+    }
+
+    /**
+     * Compares the consensus timestamp of the consensus data to the timestamp of the bound.
+     *
+     * @param consensusData the consensus data to compare to
+     * @return a negative integer, zero, or a positive integer based on whether the consensus data is less, equal or greater than the bound.
+     */
+    private int compareTimestamp(@NonNull final ConsensusData consensusData) {
+        return consensusData.getConsensusTimestamp().compareTo(timestamp);
+    }
+
+    /**
+     * Checks if the bound is unbounded.
+     *
+     * @return true if the bound is unbounded, false otherwise
+     */
+    public boolean isUnbounded() {
+        return round == NO_ROUND && timestamp == NO_TIMESTAMP;
+    }
+
+    /**
+     * The type of bound.
+     */
+    public enum BoundType {
+        LOWER,
+        UPPER
```

Comment:
I don't see this upper bound used anywhere except for unit tests... is the complexity for supporting both upper and lower bounds worth it?

Reply:
If we ever want a bounded query of the event stream (upper and lower) then it could be done with having two of these bounds.  

I don't know how to answer if it is "worth it".

Reply:
I am willing to remove the non-executing code (Upper bound) and either remove the exception blocking the use of both bounds, or remove the unreachable code if that is what is preferred.

Reply:
I have removed the non-executing code.